### PR TITLE
feat: verify OA v3

### PR DESCRIPTION
**Pull request by yehjxraymond**

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
   "compilerOptions": {
     "lib": ["es2016"],
     "target": "es5",
-    "module": "es2015",
+    "module": "CommonJS",
```
Comment (Nebulis) [inline tsconfig.json:5]: Why?

File: scripts/generate.v3.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/* eslint-disable @typescript-eslint/no-empty-function,camelcase */
```
Comment (Nebulis) [inline scripts/generate.v3.ts:1]: To different in readme

File: tsconfig.json (hunk 1 above)
Comment (yehjxraymond) [reply tsconfig.json:5]: Because ts-node will fail otherwise

File: scripts/generate.v3.ts (hunk 2 above)
Comment (yehjxraymond) [reply scripts/generate.v3.ts:1]: Whut?

File: src/verifiers/documentStatus/tokenRegistry/ethereumTokenRegistryStatus.ts [hunk 3]
```diff
@@ -42,18 +44,37 @@ export const getTokenRegistry = (
...
   }
+
+  throw new CodedError(
+    `Unrecognized document version`,
+    OpenAttestationEthereumTokenRegistryStatusCode.UNRECOGNIZED_DOCUMENT,
+    OpenAttestationEthereumTokenRegistryStatusCode[OpenAttestationEthereumTokenRegistryStatusCode.UNRECOGNIZED_DOCUMENT]
+  );
+};
+
+const getMerkleRoot = (
```
Comment (rjchow) [inline src/verifiers/documentStatus/tokenRegistry/ethereumTokenRegistryStatus.ts:68]: find it a bit strange that this isn't inside open-attestation

File: src/verifiers/issuerIdentity/dnsTxt/openAttestationDnsTxt.ts [hunk 4]
```diff
@@ -59,103 +57,141 @@ const resolveIssuerIdentity = async (
...
+    status: issuerIdentity.status,
+  };
+};
+
+const verify: VerifierType["verify"] = withCodedErrorHandler(
+  async (document, options) => {
+    if (utils.isWrappedV2Document(document)) return verifyV2(document, options);
+    if (utils.isWrappedV3Document(document)) return verifyV3(document, options);
+    throw new CodedError(
+      "Unrecognized document",
```
Comment (rjchow) [inline src/verifiers/issuerIdentity/dnsTxt/openAttestationDnsTxt.ts:176]: message is a bit vague, "Document does not match either v2 or v3 formats" ?

File: src/types/error.ts [hunk 5]
```diff
@@ -48,25 +53,33 @@ export enum OpenAttestationDidSignedDocumentStatusCode {
...
   SKIPPED = 0,
   UNEXPECTED_ERROR = 1,
   MALFORMED_IDENTITY_PROOF = 2,
   INVALID_ISSUERS = 3,
+  UNSIGNED = 4,
+  UNRECOGNIZED_DOCUMENT = 5,
 }
 export enum OpenAttestationSignatureCode {
   UNEXPECTED_ERROR = 0,
   KEY_MISSING = 1,
```
[the remark this answers is not in the document]
Comment (yehjxraymond) [reply src/types/error.ts:80]: can append, not replace

File: src/verifiers/documentStatus/tokenRegistry/ethereumTokenRegistryStatus.ts (hunk 3 above)
Comment (yehjxraymond) [reply src/verifiers/documentStatus/tokenRegistry/ethereumTokenRegistryStatus.ts:68]: yea probably can add as a util there, good idea

File: src/verifiers/issuerIdentity/dnsTxt/openAttestationDnsTxt.ts (hunk 4 above)
Comment (yehjxraymond) [reply src/verifiers/issuerIdentity/dnsTxt/openAttestationDnsTxt.ts]: done

File: src/verifiers/documentStatus/documentStore/ethereumDocumentStoreStatus.ts [hunk 6]
```diff
@@ -62,16 +63,7 @@ export const getIssuersDocumentStores = (
...
-  if (proof.method !== "DOCUMENT_STORE")
-    throw new CodedError(
-      `Document not issued using document store`,
-      OpenAttestationEthereumDocumentStoreStatusCode.INVALID_VALIDATION_METHOD,
-      OpenAttestationEthereumDocumentStoreStatusCode[
-        OpenAttestationEthereumDocumentStoreStatusCode.INVALID_VALIDATION_METHOD
-      ]
-    );
-  return [proof.value];
+  throw new Error("TBD");
```
[the remark this answers is not in the document]
Comment (yehjxraymond) [reply src/verifiers/documentStatus/documentStore/ethereumDocumentStoreStatus.ts]: removed unreachable code path

File: scripts/generate.v3.ts [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+  timeout: 5 * 60 * 1000, // 5 min timeout for contract to execute
+};
+
+const generateDnsDid = async () => {
+  info("Generating DNS-DID files");
+  writeFileSync("./test/fixtures/v3/dnsdid.json", JSON.stringify(baseDnsDidDocument, null, 2));
+  const wrappedBaseDnsDidDocument = await __unsafe__use__it__at__your__own__risks__wrapDocument(baseDnsDidDocument);
+  writeFileSync("./test/fixtures/v3/dnsdid-wrapped.json", JSON.stringify(wrappedBaseDnsDidDocument, null, 2));
+  const { merkleRoot } = wrappedBaseDnsDidDocument.proof;
+  const signature = await wallet.signMessage(utils.arrayify(`0x${merkleRoot}`));
```
[the remark this answers is not in the document]
Comment (yehjxraymond) [reply scripts/generate.v3.ts]: done

File: scripts/generate.v3.ts (hunk 2 above)
Comment (yehjxraymond) [reply scripts/generate.v3.ts:1]: done

File: src/verifiers/documentStatus/tokenRegistry/ethereumTokenRegistryStatus.ts (hunk 3 above)
Comment (yehjxraymond) [reply src/verifiers/documentStatus/tokenRegistry/ethereumTokenRegistryStatus.ts:68]: https://www.pivotaltracker.com/story/show/176860629

**Review by rjchow — APPROVED** [review]

**Pull request merged**